### ISSUE TITLE
Question: Further Documentation for agent

### ISSUE THREAD

**Issue by ASchwad** [AbsaOSS/spline-spark-agent#124]
Is there some documentation upon the data model in Arango DB and the internals of the agent? 

I am curious about your design decisions and requirements for the existing backend model. According to the spline paper, there is a mongoDB running under the hood, which seems a bit confusing. 
I am asking, because we are thinking about connecting the agent to AWS Neptune. I could also imagine, that having a more detailed architecture overview / documentation, could enable more collaborators on this project.

Last but not least: Thanks for this great project.

**Comment by wajda** [comment]
Hi @ASchwad,
Yes, you are right. We need to write some more detailed developer oriented doc eventually. We'll try to do it as soon as we complete our a few outstanding big features (like attribute level end-to-end lineage or streaming support) that might have a severe impact on the Spline internals. When our design decisions are proved to be sufficient to accommodate those requirements we'll consider the architecture stable enough and worth being documented in details. And we'll make it a major GA release 1.0.0
For now just keep asking and we'll be answering questions.

Regarding the paper, it's dated. Starting from Spline 0.4 we significantly revised the architecture, introduced a server-side module and replaced MongoDB with ArangoDB. We've also expanded our vision, and now we're moving Spline toward being more of a general data lineage tracking tool, capable to cover different possible data frameworks and pipeline used in an arbitrary organization, not only focusing on Spark like described in the paper.

Currently we have two data models:
1. The one called _Producer model_ that is used by the agent to send the data to the Spine server over the REST API.
2. The backend persistent model. 

The backend persistent data model is a combination of graph-based and document-based models (that's why we switched to ArangoDB). Spline captures execution plans that consist of smaller entities describing static aspects of the data transformation modeled as documents (like _operation_, _data source_, _attribute_, _expression_ etc). It also captures the facts of execution (we call it _execution event_ or sometimes _progress_) that holds timestamp info and bind everything together in a giant multi-dimensional graph that can later be queried in all possible ways to get different views and perspectives on the lineage data, that suites the consumer needs the best. We are planning to have a flexible query API (we call it _Consumer API_), as well as a mechanism to easily export data (or a query result) to external sources (e.g. HDFS, ElasticSearch etc) in document or graph formats for further external analysis.

_Producer API_ is versioned and considered more stable than the _Consumer API_ which is deep under development and is a subject for braking changes without notice. For _Producer API_ on the other hand, to avoid the gaps in data lineage and make the Spline upgrade process as smooth as possible in production, we try to maintain backward compatibility for a reasonable time frame. So that the older agents used across the corporate network can continue working after the Spline server upgrade and being upgraded gradually and smoothly later on. 

Hope it helps.
